src/Storages/MergeTree/MergeTreeIndexSet.cpp

```diff
@@ -318,17 +333,214 @@ bool MergeTreeIndexConditionSet::mayBeTrueOnGranule(MergeTreeIndexGranulePtr idx
+    const ContextPtr & context,
+    std::unordered_map<const ActionsDAG::Node *, const ActionsDAG::Node *> & node_to_result_node) const
+{
+    auto result_node_it = node_to_result_node.find(&node);
+    if (result_node_it != node_to_result_node.end())
+        return *result_node_it->second;
+
+    const ActionsDAG::Node * result_node = nullptr;
+
+    if (const auto * operator_node_ptr = operatorFromDAG(node, result_dag, context, node_to_result_node))
+    {
+        result_node = operator_node_ptr;
+    }
+    else if (const auto * atom_node_ptr = atomFromDAG(node, result_dag, context))
+    {
+        result_node = atom_node_ptr;
+
+        if (atom_node_ptr->type == ActionsDAG::ActionType::INPUT ||
+            atom_node_ptr->type == ActionsDAG::ActionType::FUNCTION)
+        {
+            auto bit_wrapper_function = FunctionFactory::instance().get("__bitWrapperFunc", context);
+            result_node = &result_dag->addFunction(bit_wrapper_function, {atom_node_ptr}, {});
+        }
+    }
+    else
+    {
+        ColumnWithTypeAndName unknown_field_column_with_type;
+
+        unknown_field_column_with_type.name = calculateConstantActionNodeName(UNKNOWN_FIELD);
+        unknown_field_column_with_type.type = std::make_shared<DataTypeUInt8>();
+        unknown_field_column_with_type.column = unknown_field_column_with_type.type->createColumnConst(1, UNKNOWN_FIELD);
+
+        result_node = &result_dag->addColumn(unknown_field_column_with_type);
+    }
+
+    node_to_result_node.emplace(&node, result_node);
+    return *result_node;
+}
+
+const ActionsDAG::Node * MergeTreeIndexConditionSet::atomFromDAG(const ActionsDAG::Node & node, ActionsDAGPtr & result_dag, const ContextPtr & context) const
+{
+    /// Function, literal or column
+
+    const auto * node_to_check = &node;
+    while (node_to_check->type == ActionsDAG::ActionType::ALIAS)
+        node_to_check = node_to_check->children[0];
+
+    if (node_to_check->column && isColumnConst(*node_to_check->column))
+        return &node;
+
+    RPNBuilderTreeContext tree_context(context);
+    RPNBuilderTreeNode tree_node(node_to_check, tree_context);
+
+    auto column_name = tree_node.getColumnName();
+    if (key_columns.contains(column_name))
+    {
+        const auto * result_node = node_to_check;
+
+        if (node.type != ActionsDAG::ActionType::INPUT)
+            result_node = &result_dag->addInput(column_name, node.result_type);
+
+        return result_node;
+    }
+
+    if (node.type != ActionsDAG::ActionType::FUNCTION)
+        return nullptr;
+
+    const auto & arguments = node.children;
+    size_t arguments_size = arguments.size();
+
+    ActionsDAG::NodeRawConstPtrs children(arguments_size);
+
+    for (size_t i = 0; i < arguments_size; ++i)
+    {
+        children[i] = atomFromDAG(*arguments[i], result_dag, context);
+
+        if (!children[i])
+            return nullptr;
+    }
+
+    return &result_dag->addFunction(node.function_builder, children, {});
```

Comment:
Do I understand correctly, that here we just "clone" the function node, because argument types are not changed? That's important in terms of #43761

Reply:
Yes. Argument types must be the same.